java-checks/src/main/java/org/sonar/java/checks/BatchSQLStatementsCheck.java

```diff
@@ -0,0 +1,121 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.EnumSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import static org.sonar.java.checks.helpers.TreeHelper.findClosestParentOfKind;
+
+/**
+ *
+ */
+@Rule(key = "S6912")
+public class BatchSQLStatementsCheck extends IssuableSubscriptionVisitor {
+  private static final String MESSAGE = "Use \"addBatch\" and \"executeBatch\" to execute multiple SQL statements in a single call.";
+  private static final MethodMatchers EXECUTE_METHODS = MethodMatchers.create()
+    .ofSubTypes("java.sql.Statement")
+    .names("execute", "executeQuery", "executeUpdate")
+    .withAnyParameters()
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    var methodTree = (MethodTree) tree;
+
+    var invocationsCollector = new StatementExecuteInvocationsCollector();
+    methodTree.accept(invocationsCollector);
+
+    var inLambdaInvocationsCollector = new InLambdaStatementExecuteInvocationsCollector();
+    methodTree.accept(inLambdaInvocationsCollector);
+
+    if (invocationsCollector.shouldReport() || inLambdaInvocationsCollector.shouldReport()) {
+      invocationsCollector.invocations().forEach(invocation -> reportIssue(invocation, MESSAGE));
+      inLambdaInvocationsCollector.invocations().forEach(invocation -> reportIssue(invocation, MESSAGE));
+    }
+  }
+
+  private static class StatementExecuteInvocationsCollector extends BaseTreeVisitor {
+    private static final Set<Tree.Kind> LOOP_KINDS = EnumSet.of(
+      Tree.Kind.FOR_STATEMENT,
+      Tree.Kind.FOR_EACH_STATEMENT,
+      Tree.Kind.WHILE_STATEMENT,
+      Tree.Kind.DO_STATEMENT);
+    private final List<MethodInvocationTree> invocations = new LinkedList<>();
```

Comment:
Why a LinkedList here? In most scenarios an ArrayList turns out to be more efficient.

Reply:
The linked list is more efficient for append-only operations, like in this case. The reading operations are on the whole list and not in random access, so it should be a better choice here.